client/allocdir/task_dir.go

```diff
@@ -145,12 +170,33 @@ func (t *TaskDir) Build(createChroot bool, chroot map[string]string) error {
 	}
 
 	// Build chroot if chroot filesystem isolation is going to be used
-	if createChroot {
+	if fsi == fsisolation.Chroot {
 		if err := t.buildChroot(chroot); err != nil {
 			return err
 		}
 	}
 
+	// Only bind mount the task alloc/task dirs to the client.mounts_dir/<task>
+	if fsi == fsisolation.Unveil {
+		uid, gid, _, err := dynamic.LookupUser(username)
+		if err != nil {
+			return fmt.Errorf("Failed to lookup user: %v", err)
+		}
+
+		// create the task unique directory under the client mounts path
+		parent := filepath.Dir(t.MountsAllocDir)
```

Comment:
If I'm understanding correctly, we're:
* creating `$datadir/mounts/$allocID-$task` with permissions 0710
* bind-mounting from `$datadir/allocs/$allocID` to `$datadir/mounts/$allocID-$task/alloc` with permissions 0710
* bind-mounting from `$datadir/allocs/$allocID/$task` to `$datadir/mounts/$allocID-$task/task` with permissions 0710

If so, that all seems ok... but where is `$datadir/mounts` getting created with the recommended permissions of 0700? I seem to remember that we don't set `$datadir/allocs` with the permissions we'd really recommend but now it's a backcompat issue, so if we could ensure we're creating the mounts dir with the right permissions the first time we could probably avoid some future headaches.

Reply:
Ah dangit I [that part](https://github.com/hashicorp/nomad/compare/main...exec2-firstpass#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09R692-R699) out. The permissions on `$datadir/mounts` would be `0711` - the point of all this being the outside->in protections come from `$datadir/mounts/%allocID-$task` being owned by the user of the task (`0710` as mentioned above),  which itself can be a "dynamic workload user" that does not exist from the perspective of the OS. I'll add the directory creation in another commit.

Reply:
Added directory creation in ee9e854f94